### PR TITLE
RDoc-1072 documenting test driver

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+using System.Linq;
+using Raven.Client.Documents.Indexes;
+
+namespace RavenDBTestDriver
+{
+    public class RavenDBTestDriver : RavenTestDriver<MyRavenDBLocator>
+    {
+        #region test_driver_2
+        //This allows us to generate a clean database for our test
+        protected override void SetupDatabase(IDocumentStore documentStore)
```
Comment (ayende) [inline Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs]: Why do we need this? Unless I remember badly, we are already doing this for the user.

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            var doc = new DatabaseRecord(documentStore.Database);
+            try
+            {
+                documentStore.Maintenance.Server.Send(new CreateDatabaseOperation(doc));
+            }
+            catch (Exception e)
+            {
+                //Ignore if the database exists, we could also delete it and create a new one if we need
+                if (e.Message.EndsWith("already exists!") == false)
```
Comment (ayende) [inline Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs]: Let's avoid such error handling.

File: Documentation/4.0/Raven.Documentation.Pages/start/test-driver.markdown [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+{PANEL:PreInitializingTheStore}
+
+Pre-Initializing the IDocumentStore allows you to mutate the conventions used by the document store.
+### Example
+{CODE test_driver_3@Start\RavenDBTestDriver.cs /}
+
+{PANEL/}
+
+{PANEL:UnitTest}
+Finally we can write down a simple test, note that i'm using [xunit](https://www.nuget.org/packages/xunit/) for my test framework in the below example.
```
Comment (ayende) [inline Documentation/4.0/Raven.Documentation.Pages/start/test-driver.markdown]: `i'm` -> `I'm`

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriverFull.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                    var query = session.Query<TestDocument, TestDocumentByName>().Where(x => x.Name == "hello").ToList();
+                    Assert.Single(query);
+                }
+            }
+        }
+    }
+
+    public class MyRavenDBLocator : RavenServerLocator
+    {
+        public override string ServerPath => @"C:\work\RavenDB-4.0.0-rc-40025-windows-x64\Server\Raven.Server.dll";
```
Comment (ayende) [inline Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriverFull.cs]: Add some logic here, so people won't think that there is just a const here.
Something like if search a few paths and return the first that match,etc

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs (hunk 1 above)
Comment (talweiss1982) [reply Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs]: we only take care of it if paths are provided from what i saw

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs (hunk 2 above)
Comment (talweiss1982) [reply Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs]: taken from the source code of ravendb

Comment (ppekrol) [reply Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs]: @talweiss1982 I think that it should throw ConcurrencyException (or I might be wrong), we can catch that

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs (hunk 1 above)
Comment (gregolsky) [reply Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs]: We create a DB for the user already https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.TestDriver/RavenTestDriver.cs#L59
Additionally he's able to provide `DatabaseDumpFilePath` or `DatabaseDumpFileStream` if he wants it imported there.

**Comment by talweiss1982** [comment]
done

File: Documentation/4.0/Raven.Documentation.Pages/start/test-driver.markdown [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+| **protected virtual void SetupDatabase(IDocumentStore documentStore)** | Allows you to initialize the database. |
+| **protected event EventHandler DriverDisposed;** |An event that is raised when the test driver is been disposed of. |
+| **public void WaitForIndexing(IDocumentStore store, string database = null, TimeSpan? timeout = null)** | Allows you to wait for indexes to become non-stale. |
+| **public void WaitForUserToContinueTheTest(IDocumentStore store)** | Allows you to break the test and launch the studio to examine the state of the database. |
+| **protected virtual void OpenBrowser(string url)** | Allows you to open the browser. |
+| **public virtual void Dispose()** | Allows you to dispose of the server. |
+
+{PANEL/}
+
+{PANEL:PreInitializingTheStore}
```
Comment (ppekrol) [inline Documentation/4.0/Raven.Documentation.Pages/start/test-driver.markdown]: 'PreInitialize' would be enough?

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs [hunk 6]
```diff
@@ -23,10 +23,6 @@ protected override void SetupDatabase(IDocumentStore documentStore)
             catch (Exception e)
             {
                 //Ignore if the database exists, we could also delete it and create a new one if we need
-                if (e.Message.EndsWith("already exists!") == false)
```
Comment (ayende) [inline Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriver.cs]: The point of `SetupDatabase` is to allow you to do things like create a dataset.
There is no need to create db here, and this should be removed.

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriverFull.cs [hunk 7]
```diff
@@ -61,7 +57,13 @@ public void MyFirstTest()
 
     public class MyRavenDBLocator : RavenServerLocator
     {
-        public override string ServerPath => @"C:\work\RavenDB-4.0.0-rc-40025-windows-x64\Server\Raven.Server.dll";
+        public override string ServerPath
+        {
+            get
+            {
+                return Environment.GetEnvironmentVariable("RavenServerTestPath");
```
Comment (ayende) [inline Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Start/RavenDBTestDriverFull.cs]: Let's use multiple paths, I want to make it explicit that there can be logic there.
So you will search each of the paths for FileExists

**Comment by PureKrome** [comment]
Thank you!

**Pull request merged**